components/logins/src/error.rs

```diff
@@ -3,19 +3,49 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use std::ffi::OsString;
+pub type Result<T> = std::result::Result<T, LoginsError>;
+// Functions which are part of the public API should use this Result.
+pub type ApiResult<T> = std::result::Result<T, LoginsStorageError>;
+
+pub use error_support::handle_error;
+use error_support::{ErrorHandling, GetErrorHandling};
+
+// Errors we return via the public interface.
+//
+// Named `LoginsStorageError` for backwards compatibility reasons, although
+// this name shouldn't need to be used anywhere other than this file and the .udl
+//
+// Note that there is no `Into` between public and internal errors, but
+// instead the `ErrorHandling` mechanisms are used to explicitly convert
+// when necessary.
+//
+// XXX - not clear that these actually need to use `thiserror`? Certainly
+// not necessary to use `#[from]` though.
+#[derive(Debug, thiserror::Error)]
+pub enum LoginsStorageError {
+    #[error("Invalid login: {0}")]
+    InvalidRecord(String),
+
+    #[error("No record with guid exists (when one was required): {0:?}")]
+    NoSuchRecord(String),
+
+    #[error("Encryption key is in the correct format, but is not the correct key.")]
+    IncorrectKey,
+
+    #[error("{0}")]
+    Interrupted(String),
 
-// TODO: this is (IMO) useful and was dropped from `failure`, consider moving it
-// into `error_support`.
-macro_rules! throw {
-    ($e:expr) => {
-        return Err(Into::into($e))
-    };
+    #[error("Unexpected Error: {0}")]
+    UnexpectedLoginsStorageError(String),
```

Comment:
There's a couple errors that might be useful here in the short term: `SyncAuthError` and `RequestFailed`, which I think can be returned by the `sync()` method.  iOS is still using that method, and might want to do treat those error specially.  `SyncAuthError` means the app should tell the user they're not logged in.  `RequestFailed` is an expected error, so maybe that affects how iOS wants to log/report/track things.  I'm not really sure though, what do others think?